In [1]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Dev_data_to_be_shared.csv')

df.head()

account_number  bad_flag  onus_attribute_1  transaction_attribute_1  \
0               1         0               NaN                      NaN   
1               2         0          221000.0                      0.0   
2               3         0           25000.0                      0.0   
3               4         0           86000.0                      0.0   
4               5         0          215000.0                      0.0   

   transaction_attribute_2  transaction_attribute_3  transaction_attribute_4  \
0                      NaN                      NaN                      NaN   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                      0.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   transaction_attribute_5  transaction_attribute_6  transaction_attribute_7  \
0                      NaN                      NaN                      NaN   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                      0.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   ...  bureau_enquiry_47  bureau_enquiry_48  bureau_enquiry_49  \
0  ...                0.0                0.0                0.0   
1  ...                0.0                0.0                2.0   
2  ...                0.0                0.0                0.0   
3  ...                0.0                0.0                0.0   
4  ...                0.0                0.0                0.0   

   bureau_enquiry_50  onus_attribute_43  onus_attribute_44  onus_attribute_45  \
0                1.0                NaN                NaN                NaN   
1                3.0                0.0                0.0                0.0   
2                8.0                NaN                NaN                NaN   
3               30.0                NaN                NaN                NaN   
4                1.0                NaN                NaN                NaN   

   onus_attribute_46  onus_attribute_47  onus_attribute_48  
0                NaN                NaN                NaN  
1                0.0                0.0                0.0  
2                NaN                NaN                NaN  
3                NaN                NaN                NaN  
4                NaN                NaN                NaN  

[5 rows x 1216 columns]

In [2]:
# Data Overview
# Check the shape of the dataset and data types
df_shape = df.shape
df_dtypes = df.dtypes

# Display the shape and data types
df_shape, df_dtypes.head()

((96806, 1216),
 account_number               int64
 bad_flag                     int64
 onus_attribute_1           float64
 transaction_attribute_1    float64
 transaction_attribute_2    float64
 dtype: object)

The dataset contains **96,806 rows** and **1,216 columns**. The first few columns have the following data types:

- `account_number`: int64
- `bad_flag`: int64
- `onus_attribute_1`: float64
- `transaction_attribute_1`: float64
- `transaction_attribute_2`: float64

Next, let's identify columns with missing values and their proportions.

In [3]:
# Calculate the percentage of missing values in each column
missing_values = df.isnull().mean() * 100

# Filter columns with missing values
missing_values = missing_values[missing_values > 0]

# Sort by percentage of missing values
df_missing_values = missing_values.sort_values(ascending=False)

df_missing_values.head(10)

bureau_447           100.000000
bureau_436           100.000000
bureau_449            94.124331
bureau_148            93.552053
bureau_448            90.028511
onus_attribute_45     88.006942
onus_attribute_44     88.006942
onus_attribute_43     88.006942
onus_attribute_48     88.006942
onus_attribute_47     88.006942
dtype: float64

In [4]:
# Descriptive statistics for numerical columns
descriptive_stats = df.describe()
descriptive_stats

account_number      bad_flag  onus_attribute_1  \
count    96806.000000  96806.000000      7.157500e+04   
mean     48403.500000      0.014173      1.542391e+05   
std      27945.629417      0.118203      1.729925e+05   
min          1.000000      0.000000      2.500000e+04   
25%      24202.250000      0.000000      5.900000e+04   
50%      48403.500000      0.000000      1.000000e+05   
75%      72604.750000      0.000000      1.810000e+05   
max      96806.000000      1.000000      2.800000e+06   

       transaction_attribute_1  transaction_attribute_2  \
count             71575.000000             71575.000000   
mean                  9.570769                 0.002207   
std                1513.967595                 0.119240   
min                   0.000000                 0.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max              398936.791000                25.000000   

       transaction_attribute_3  transaction_attribute_4  \
count             71575.000000             71575.000000   
mean                  4.092854                77.306435   
std                 301.580599              3164.987013   
min                   0.000000           -109800.476600   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max               55000.000000            358986.000000   

       transaction_attribute_5  transaction_attribute_6  \
count             71575.000000             71575.000000   
mean                  0.006315                67.043550   
std                   0.129545              2516.330899   
min                   0.000000             -3498.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max                  12.000000            358986.000000   

       transaction_attribute_7  ...  bureau_enquiry_47  bureau_enquiry_48  \
count             71575.000000  ...            94212.0       94212.000000   
mean                  0.081034  ...                0.0           0.189657   
std                  14.203615  ...                0.0           0.597298   
min                   0.000000  ...                0.0           0.000000   
25%                   0.000000  ...                0.0           0.000000   
50%                   0.000000  ...                0.0           0.000000   
75%                   0.000000  ...                0.0           0.000000   
max                3150.000000  ...                0.0          18.000000   

       bureau_enquiry_49  bureau_enquiry_50  onus_attribute_43  \
count       94212.000000       94212.000000       11610.000000   
mean            0.044283           7.800673           0.588114   
std             0.369451           8.655149           1.059613   
min             0.000000           0.000000           0.000000   
25%             0.000000           2.000000           0.000000   
50%             0.000000           5.000000           0.000000   
75%             0.000000          11.000000           1.000000   
max            14.000000         102.000000          19.000000   

       onus_attribute_44  onus_attribute_45  onus_attribute_46  \
count       11610.000000       11610.000000       11610.000000   
mean            1.497158           0.545650           1.430491   
std             1.702795           1.055865           1.728950   
min             0.000000           0.000000           0.000000   
25%             1.000000           0.000000           1.000000   
50%             1.000000           0.000000           1.000000   
75%             2.000000           1.000000           2.000000   
max            38.000000          19.000000          38.000000   

       onus_a

In [5]:
# Calculate correlation matrix
correlation_matrix = df.corr()

# Extract correlation of `bad_flag` with other features
bad_flag_correlation = correlation_matrix['bad_flag'].sort_values(ascending=False)

# Display the top 10 features most correlated with `bad_flag`
bad_flag_correlation.head(10)

bad_flag             1.000000
onus_attribute_2     0.108491
onus_attribute_17    0.103156
onus_attribute_23    0.100748
onus_attribute_20    0.100672
onus_attribute_26    0.098549
onus_attribute_32    0.096320
onus_attribute_29    0.093453
onus_attribute_33    0.092944
onus_attribute_35    0.090654
Name: bad_flag, dtype: float64

In [6]:
# Feature Engineering
# Example: Create a new feature that is the sum of some of the onus attributes
# This is just an example; in practice, you would create features based on domain knowledge

df['onus_sum'] = df[['onus_attribute_1', 'onus_attribute_2', 'onus_attribute_3']].sum(axis=1)

# Check the new feature
new_feature_stats = df['onus_sum'].describe()
new_feature_stats

count    9.680600e+04
mean     1.140393e+05
std      1.634346e+05
min     -8.487496e-01
25%      9.211882e-01
50%      6.900000e+04
75%      1.450000e+05
max      2.800001e+06
Name: onus_sum, dtype: float64

In [4]:
import pandas as pd

# Reload the dataset
df = pd.read_csv('Dev_data_to_be_shared.csv')

df.head()

account_number  bad_flag  onus_attribute_1  transaction_attribute_1  \
0               1         0               NaN                      NaN   
1               2         0          221000.0                      0.0   
2               3         0           25000.0                      0.0   
3               4         0           86000.0                      0.0   
4               5         0          215000.0                      0.0   

   transaction_attribute_2  transaction_attribute_3  transaction_attribute_4  \
0                      NaN                      NaN                      NaN   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                      0.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   transaction_attribute_5  transaction_attribute_6  transaction_attribute_7  \
0                      NaN                      NaN                      NaN   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                      0.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   ...  bureau_enquiry_47  bureau_enquiry_48  bureau_enquiry_49  \
0  ...                0.0                0.0                0.0   
1  ...                0.0                0.0                2.0   
2  ...                0.0                0.0                0.0   
3  ...                0.0                0.0                0.0   
4  ...                0.0                0.0                0.0   

   bureau_enquiry_50  onus_attribute_43  onus_attribute_44  onus_attribute_45  \
0                1.0                NaN                NaN                NaN   
1                3.0                0.0                0.0                0.0   
2                8.0                NaN                NaN                NaN   
3               30.0                NaN                NaN                NaN   
4                1.0                NaN                NaN                NaN   

   onus_attribute_46  onus_attribute_47  onus_attribute_48  
0                NaN                NaN                NaN  
1                0.0                0.0                0.0  
2                NaN                NaN                NaN  
3                NaN                NaN                NaN  
4                NaN                NaN                NaN  

[5 rows x 1216 columns]

In [5]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
import xgboost as xgb

# Prepare the data
X = df.drop(columns=['bad_flag'])
y = df['bad_flag']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Define the XGBoost model
xgb_model = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss')

# Fit the model
xgb_model.fit(X_train, y_train)

# Predict on the test set
y_pred = xgb_model.predict(X_test)

# Calculate accuracy and classification report
accuracy = accuracy_score(y_test, y_pred)
classification_report_str = classification_report(y_test, y_pred)

accuracy, classification_report_str

/usr/local/lib/python3.11/site-packages/xgboost/core.py:158: UserWarning: [16:09:55] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)


(0.9859002169197397,
 '              precision    recall  f1-score   support\n\n           0       0.99      1.00      0.99     19092\n           1       0.29      0.01      0.01       270\n\n    accuracy                           0.99     19362\n   macro avg       0.64      0.50      0.50     19362\nweighted avg       0.98      0.99      0.98     19362\n')

In [6]:
from catboost import CatBoostClassifier

# Define the CatBoost model
catboost_model = CatBoostClassifier(verbose=0, random_seed=42)

# Fit the model
catboost_model.fit(X_train, y_train)

# Predict on the test set
y_pred_catboost = catboost_model.predict(X_test)

# Calculate accuracy and classification report
accuracy_catboost = accuracy_score(y_test, y_pred_catboost)
classification_report_catboost = classification_report(y_test, y_pred_catboost)

accuracy_catboost, classification_report_catboost

ModuleNotFoundError: No module named 'catboost'

In [7]:
!pip install catboost

Defaulting to user installation because normal site-packages is not writeable


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0.0/98.7 MB ? eta -:--:--

   ━━╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.9/98.7 MB 100.8 MB/s eta 0:00:01

   ━━━━━╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.6/98.7 MB 232.4 MB/s eta 0:00:01

   ━━━━━━━━━╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.5/98.7 MB 132.3 MB/s eta 0:00:01

   ━━━━━━━━━━━━╺━━━━━━━━━━━━━━━━━━━━━━━━━━━ 30.4/98.7 MB 169.4 MB/s eta 0:00:01

   ━━━━━━━━━━━━╺━━━━━━━━━━━━━━━━━━━━━━━━━━━ 30.4/98.7 MB 169.4 MB/s eta 0:00:01

   ━━━━━━━━━━━━╸━━━━━━━━━━━━━━━━━━━━━━━━━━━ 31.0/98.7 MB 62.6 MB/s eta 0:00:02

   ━━━━━━━━━━━━━╺━━━━━━━━━━━━━━━━━━━━━━━━━━ 32.4/98.7 MB 62.5 MB/s eta 0:00:02

   ━━━━━━━━━━━━━━━━━━━╺━━━━━━━━━━━━━━━━━━━━ 47.7/98.7 MB 212.7 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━╺━━━━━━━━━━━━━━━━━ 55.5/98.7 MB 225.1 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸━━━━━━━━━━━ 70.9/98.7 MB 218.5 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╺━━━━━━━━ 76.5/98.7 MB 180.0 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸━━━ 91.2/98.7 MB 222.6 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸ 98.7/98.7 MB 226.5 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸ 98.7/98.7 MB 226.5 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸ 98.7/98.7 MB 226.5 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸ 98.7/98.7 MB 226.5 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸ 98.7/98.7 MB 226.5 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸ 98.7/98.7 MB 226.5 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸ 98.7/98.7 MB 226.5 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸ 98.7/98.7 MB 226.5 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸ 98.7/98.7 MB 226.5 MB/s eta 0:00:01

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.7/98.7 MB 21.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.1/47.1 kB 6.0 MB/s eta 0:00:00



[notice] A new release of pip is available: 24.0 -> 24.3.1
[notice] To update, run: pip install --upgrade pip


In [8]:
from catboost import CatBoostClassifier

# Define the CatBoost model
catboost_model = CatBoostClassifier(verbose=0)

# Fit the model
catboost_model.fit(X_train, y_train)

# Predict on the test set
y_pred = catboost_model.predict(X_test)

# Calculate accuracy and classification report
accuracy = accuracy_score(y_test, y_pred)
classification_report_str = classification_report(y_test, y_pred)

accuracy, classification_report_str

(0.9860035120338808,
 '              precision    recall  f1-score   support\n\n           0       0.99      1.00      0.99     19092\n           1       0.00      0.00      0.00       270\n\n    accuracy                           0.99     19362\n   macro avg       0.49      0.50      0.50     19362\nweighted avg       0.97      0.99      0.98     19362\n')